node/full_client_test.go

```diff
@@ -742,40 +770,81 @@ func TestValidatorSetHandling(t *testing.T) {
 		vals, err := rpc.Validators(context.Background(), &h, nil, nil)
 		assert.NoError(err)
 		assert.NotNil(vals)
-		assert.EqualValues(len(genesisValidators), vals.Total)
-		assert.Len(vals.Validators, len(genesisValidators))
+		assert.EqualValues(numNodes, vals.Total)
+		assert.Len(vals.Validators, numNodes)
 		assert.EqualValues(vals.BlockHeight, h)
 	}
 
 	// check for "latest block"
 	vals, err := rpc.Validators(context.Background(), nil, nil, nil)
 	assert.NoError(err)
 	assert.NotNil(vals)
-	assert.EqualValues(len(genesisValidators), vals.Total)
-	assert.Len(vals.Validators, len(genesisValidators))
+	assert.EqualValues(numNodes, vals.Total)
+	assert.Len(vals.Validators, numNodes)
 	assert.GreaterOrEqual(vals.BlockHeight, int64(9))
 }
 
-func createApp(keyToRemove tmcrypto.PrivKey, waitCh chan interface{}, require *require.Assertions) *mocks.Application {
-	app := &mocks.Application{}
-	app.On("InitChain", mock.Anything).Return(abci.ResponseInitChain{})
-	app.On("CheckTx", mock.Anything).Return(abci.ResponseCheckTx{})
-	app.On("BeginBlock", mock.Anything).Return(abci.ResponseBeginBlock{})
-	app.On("Commit", mock.Anything).Return(abci.ResponseCommit{})
-	app.On("GetAppHash", mock.Anything).Return(abci.ResponseGetAppHash{})
-	app.On("GenerateFraudProof", mock.Anything).Return(abci.ResponseGenerateFraudProof{})
+// Tests moving from a centralized validator to empty validator set
+func TestValidatorSetHandlingBased(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
 
-	pbValKey, err := encoding.PubKeyToProto(keyToRemove.PubKey())
-	require.NoError(err)
+	waitCh := make(chan interface{})
 
-	app.On("EndBlock", mock.Anything).Return(abci.ResponseEndBlock{}).Times(2)
-	app.On("EndBlock", mock.Anything).Return(abci.ResponseEndBlock{ValidatorUpdates: []abci.ValidatorUpdate{{PubKey: pbValKey, Power: 0}}}).Once()
-	app.On("EndBlock", mock.Anything).Return(abci.ResponseEndBlock{}).Once()
-	app.On("EndBlock", mock.Anything).Return(abci.ResponseEndBlock{ValidatorUpdates: []abci.ValidatorUpdate{{PubKey: pbValKey, Power: 100}}}).Once()
-	app.On("EndBlock", mock.Anything).Return(abci.ResponseEndBlock{}).Run(func(args mock.Arguments) {
-		waitCh <- nil
-	})
-	return app
+	numNodes := 1
```

Comment:
[non-blocking][proposal]: Additional scenarios that covers #836 

- [ ] From N to 0 back to 1 - expected to fail
- [ ] From N to 1. Then from 1 to 2 and straight to 0

Reply:
The first scenario of N to 0 and back to 1 - expected to fail is handled in #854. For the second, created a separate issue #863